src/cluster.c

```diff
@@ -4948,6 +4986,122 @@ void addNodeReplyForClusterSlot(client *c, clusterNode *node, int start_slot, in
+    reply_cnt++;
+    if (node->hostname) {
+        addReplyBulkCString(c, "hostname");
+        addReplyBulkCString(c, node->hostname);
+        reply_cnt++;
+    }
+    addReplyBulkCString(c, "role");
+    if (nodeIsSlave(node)) {
+        addReplyBulkCString(c, "replica");
+        addReplyBulkCString(c, "replication-offset");
+        addReplyLongLong(c, node->repl_offset);
+        reply_cnt+=2;
+    } else {
+        addReplyBulkCString(c, "master");
+        reply_cnt++;
+    }
+    addReplyBulkCString(c, "health");
+    const char *health_msg = NULL;
+    if (nodeFailed(node)) {
+        health_msg = "FAIL";
+    } else if (nodeIsSlave(node) && node->repl_offset == 0) {
+        health_msg = "LOADING";
+    } else {
+        health_msg = "ONLINE";
+    }
+    addReplyBulkCString(c, health_msg);
+    reply_cnt++;
+
+    setDeferredMapLen(c, node_replylen, reply_cnt);
+}
+
+void addNodeReplyForClusterShard(client *c, clusterNode *node, list *slot_info_pair) {
+    addReplyMapLen(c, 2);
+    addReplyBulkCString(c,"slots");
+    uint16_t slot_pair_count = 0;
+    if (slot_info_pair != NULL) {
+        slot_pair_count = listLength(slot_info_pair);
+        serverAssert((slot_pair_count % 2) == 0);
+        addReplyArrayLen(c, slot_pair_count);
+        listIter li;
+        listRewind(slot_info_pair, &li);
+        listNode *ln;
+        while((ln = listNext(&li))) {
+            addReplyBulkLongLong(c, (unsigned long)listNodeValue(ln));
+            ln = listNext(&li);
+            addReplyBulkLongLong(c, (unsigned long)listNodeValue(ln));
+        }
+    } else {
+        addReplyArrayLen(c, slot_pair_count);
+    }
+    addReplyBulkCString(c,"nodes");
+    list *nodes_for_slot = clusterGetNodesServingMySlots(node);
+    if (nodes_for_slot != NULL) {
+        addReplyArrayLen(c, listLength(nodes_for_slot));
+        if (listLength(nodes_for_slot) != 0) {
+            listIter li;
+            listNode *ln;
+            listRewind(nodes_for_slot, &li);
+            while ((ln = listNext(&li))) {
+                clusterNode *node = listNodeValue(ln);
+                addNodeDetailsToShardReply(c, node);
+            }
+            listRelease(nodes_for_slot);
+        }
+    } else {
+        addReplyArrayLen(c, 0);
+    }
+}
+
+void clusterReplyShards(client *c) {
+    void *shard_replylen = addReplyDeferredLen(c);
+    int shard_cnt = 0;
+    clusterGenNodesSlotsInfo(0);
+    dictIterator *di = dictGetSafeIterator(server.cluster->nodes);
+    dictEntry *de;
+    while((de = dictNext(di)) != NULL) {
+        clusterNode *n = dictGetVal(de);
+        if (nodeIsSlave(n)) {
+            continue;
+        }
+        shard_cnt++;
+        if (n->slot_info_pair) {
+            addNodeReplyForClusterShard(c, n, n->slot_info_pair);
+            listRelease(n->slot_info_pair);
+            n->slot_info_pair = NULL;
+        } else {
+            addNodeReplyForClusterShard(c, n, NULL);
+        }
```

Comment:
I assume that the assumption is that when n->slot_info_pair is NULL, the node owns no slots? If so, let's document that.
```suggestion
        addNodeReplyForClusterShard(c, n, n->slot_info_pair);
        if (n->slot_info_pair) {
            listRelease(n->slot_info_pair);
            n->slot_info_pair = NULL;
        }
```